Adiciona CRUD de municipes

PedroPedrassolli opened the pull request:
Esse PR traz várias coisas para que a base da aplicação funcione, incluindo:

- ActiveStorage para as fotos dos municipes.
- Devise para login como usuário.
- Bootstrap com Jquery e Popper como dependências.
- Pry no lugar de Byebug para debugar.
- Rubocop para analisar o código e apontar inconsistências.
- RSpec para se criar e rodar testes.
- Decorator para as views.
- SimpleCov para gerar coverage dos testes feitos pelo RSpec.


E o básico do CRUD:

- Rotas.
- Controller.
- Model com validações.
- Views com o I18n e InputMask.
- Flash messages

Falta fazer:
- Adicionar testes, coverage atual: 90.91% ( Irei adicionar em um próximo PR, como dependência técnica )

PedroPedrassolli committed:
build(active storage): install active storage and runs migration

PedroPedrassolli committed:
build(devise): installs devise

also adds an i18n pt-br file adn switches layouts/application to application/index

PedroPedrassolli committed:
build(bootstrap): install bootstrap, jquery and popper.js using yarn

PedroPedrassolli committed:
refactor(application.js): removes bootstrap import that wasnt being used

PedroPedrassolli committed:
feat(municipe): adiciona o CRUD de municipe

falta flash messages, testes, botão de editar na table inicial e transformar o booleano de estado em
ativo e inativo ao invés de true or false

PedroPedrassolli committed:
build(pry/rubocop): instala rubocop e troca byebug por pry

também roda rubocop -a

PedroPedrassolli committed:
build(rspec): installs rspec

also adds the municipe model spec file

ricardopacheco commented on Gemfile:
```diff
@@ -29,8 +31,8 @@ gem 'jbuilder', '~> 2.7'
 group :development, :test do
-  # Call 'byebug' anywhere in the code to stop execution and get a debugger console
-  gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
+  gem 'pry'
```
Segue o padrão que vc mesmo tá usando :trollface: 

Especifica cada dependência com a versão específica!

ricardopacheco commented on Gemfile:
```diff
@@ -40,8 +42,9 @@ group :development do
   # Can be configured to work on production as well see: https://github.com/MiniProfiler/rack-mini-profiler/blob/master/README.md
   gem 'rack-mini-profiler', '~> 2.0'
   # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
+  gem 'rubocop'
```
Mesmo comentário acima!

ricardopacheco commented on Gemfile:
```diff
@@ -40,8 +42,9 @@ group :development do
 
 # Windows does not include zoneinfo files, so bundle the tzinfo-data gem
-gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
+gem 'tzinfo-data', platforms: %i[mingw mswin x64_mingw jruby]
```
Nice! new array syntax 👍

PedroPedrassolli replied:
Rubocop é companheiro 🎊

ricardopacheco commented on app/controllers/municipes_controller.rb:
```diff
@@ -0,0 +1,42 @@
+    end
+  end
+
+  def show; end
```
Excelente! Seguiu as diretivas do rubocop 👍

PedroPedrassolli replied:
Esse eu juro que adicionei antes do Rubocop, juro juradinho 😳

ricardopacheco commented on app/controllers/municipes_controller.rb:
```diff
@@ -0,0 +1,42 @@
+  private
+
+  def set_municipe
+    @municipe = Municipe.find(params[:id])
```
Sugestão: O que acontece quando params[:id] é em branco ou não encontra na base de dados?

Aqui dá pra fazer algo legal, mas, deixo ao seu critério.

PedroPedrassolli replied:
🤔 Que tal um `find_or_initialize_by` ?

PedroPedrassolli replied:
ou seria mais bacana redirecionar para uma página 404 direto?

ricardopacheco commented on app/models/municipe.rb:
```diff
@@ -0,0 +1,11 @@
+  has_one_attached :photo
+
+  validates :name, :cpf, :cns, :email, :birthdate, :phone, :photo, :status, presence: true
+  validates :cpf, length: { is: 11 }
```
Como você está usando Devise.email_regexp para validar o formato de email, fazer a validação dos campos `cpf` e `phone` podem ser feitas também por thrird dependency.

Recomendo as gems [cpf_cnpj](https://github.com/fnando/cpf_cnpj) e [phonelib](https://github.com/daddyz/phonelib).

PedroPedrassolli replied:
Opa, bacana! Vou ver de usar.

ricardopacheco commented on app/models/user.rb:
```diff
@@ -0,0 +1,4 @@
+class User < ApplicationRecord
+  devise :database_authenticatable, :registerable,
```
Será necessário todos esses módulos do devise? Lembre que quanto mais coisas você usa, mais você precisa garantir o comportamento delas, logo, é mais trabalho. Eu digo isso porque devemos ter cuidado com as sugestões "default".

Minha recomendação: Vai no GH do devise, e, leia sobre o que cada módulo dele faz. O que você ganha quando insere eles, quais limitações, etc. Assim, vc pode enxugar tendo apenas o que de fato vc precisa

PedroPedrassolli replied:
Então, esse foi o adicionado por padrão.
Irei olhar mais a fundo pra ver quais vou realmente usar e remover o resto (aposto que afeta as views criadas por padrão também, quem sabe não rola uma limpeza)

ricardopacheco commented on app/views/devise/confirmations/new.html.erb:
```diff
@@ -0,0 +1,16 @@
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Resend confirmation instructions" %>
```
I18n message.

ricardopacheco commented on app/views/devise/passwords/edit.html.erb:
```diff
@@ -0,0 +1,25 @@
+<h2>Change your password</h2>
```
I18n message onde é necessário em todo esse arquivo.

ricardopacheco commented on app/views/devise/passwords/new.html.erb:
```diff
@@ -0,0 +1,16 @@
+<h2>Forgot your password?</h2>
```
Mesmo comentário acima (I18n).

ricardopacheco commented on app/views/devise/registrations/edit.html.erb:
```diff
@@ -0,0 +1,43 @@
+<h2>Edit <%= resource_name.to_s.humanize %></h2>
```
Mesmo comentário acima (I18n).

ricardopacheco commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,29 @@
+<h2>Sign up</h2>
```
Mesmo comentário acima (I18n).

ricardopacheco commented on app/views/devise/sessions/new.html.erb:
```diff
@@ -0,0 +1,26 @@
+<h2>Log in</h2>
```
Mesmo comentário acima (I18n).

ricardopacheco commented on app/views/devise/shared/_links.html.erb:
```diff
@@ -0,0 +1,25 @@
+<%- if controller_name != 'sessions' %>
```
Mesmo comentário acima (I18n).

ricardopacheco commented on app/views/devise/unlocks/new.html.erb:
```diff
@@ -0,0 +1,16 @@
+<h2>Resend unlock instructions</h2>
```
Mesmo comentário acima (I18n).

ricardopacheco commented on app/views/layouts/application.html.erb:
```diff
@@ -1,7 +1,7 @@
 <html>
   <head>
-    <title>MunicipesApp</title>
+    <title><%= I18n.t("global.title")  %></title>
```
Excelente! 👍

ricardopacheco replied:
Excelente!

ricardopacheco commented on app/views/layouts/application.html.erb:
```diff
@@ -1,7 +1,7 @@
+    <title><%= I18n.t("global.title")  %></title>
     <meta name="viewport" content="width=device-width,initial-scale=1">
     <%= csrf_meta_tags %>
     <%= csp_meta_tag %>
```
Você sabe o que esses dois meta tags fazem? Acredito que TALVEZ um deles é desnecessário. Qual seria e porque?

ricardopacheco commented on app/views/municipes/index.html.erb:
```diff
@@ -0,0 +1,27 @@
+<table class="table table-striped table-bordered table-hover">
```
Beaufiful! <3

ricardopacheco commented on app/views/shared/_navbar.html.erb:
```diff
@@ -0,0 +1,22 @@
+      </ul>
+      <form class="d-flex">
+        <input class="form-control me-2" type="search" placeholder="Search" aria-label="Search">
+        <button class="btn btn-outline-success" type="submit">Search</button>
```
I18n.

ricardopacheco commented on bin/bundle:
```diff
@@ -8,7 +8,7 @@
 # this file is here to facilitate running it.
```
Dica geral: Arquivos em certos diretórios são padrão do próprio framework. No sentido correto, vc fez bem em alterar conforme sugestão do rubocop. Mas, geralmente isso não é recomendado no sentido prático visto que quando vc precisa reescrever esses arquivos através da atualização de versões rails major, vc acaba tendo mais problema do que solução (poiis as próprias versões novas do rails reescrevem esses arquivos).

Apesar de correto, recomendo tu configurar o rubocop para ignorar esse diretório (bem como alguns outros que sofrem do problema que falei acima).

ricardopacheco commented on config/application.rb:
```diff
@@ -1,18 +1,18 @@
 
-require "rails"
+require 'rails'
 # Pick the frameworks you want:
```
Tem algum muito interessante que você pode fazer aqui. Quero que descubra o quê.

Dica: Falei algo similar do devise acima ...

ricardopacheco commented on config/application.rb:
```diff
@@ -24,6 +24,8 @@ class Application < Rails::Application
     # Initialize configuration defaults for originally generated Rails version.
     config.load_defaults 6.1
 
+    config.i18n.default_locale = :"pt-BR"
```
Aproveita que já se molhou e dança na chuva! Configura o fallback o/

ricardopacheco commented on config/environments/development.rb:
```diff
@@ -34,6 +34,7 @@
   config.active_storage.service = :local
 
   # Don't care if the mailer can't send.
+  config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }
```
Correto, porém, não é a prática ideal. Quero que leia sobre "rails best practices configuration files"

ricardopacheco commented on config/environments/production.rb:
```diff
@@ -53,12 +53,11 @@
 
   # Prepend all log lines with the following tags.
-  config.log_tags = [ :request_id ]
+  config.log_tags = [:request_id]
```
Não estamos usando o formato de new array syntax? rs

ricardopacheco commented on config/initializers/assets.rb:
```diff
@@ -11,4 +11,4 @@
 # application.js, application.css, and all non-JS/CSS in the app/assets
 # folder are already added.
-# Rails.application.config.assets.precompile += %w( admin.js admin.css )
+Rails.application.config.assets.precompile += %w[municipes/form.js]
```
Porque a necessidade disso?

PedroPedrassolli replied:
Tinha sido erro meu, já mudei o arquivo de lugar e tirei o precompile.

ricardopacheco commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,311 @@
+  # Configure the e-mail address which will be shown in Devise::Mailer,
+  # note that it will be overwritten if you use your own mailer class
+  # with default "from" parameter.
+  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
```
Isso é importante! Dá uma lida e entenda o que isso faz!

ricardopacheco commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,311 @@
+  # confirmation, reset password and unlock tokens in the database.
+  # Devise will use the `secret_key_base` as its `secret_key`
+  # by default. You can change it below and use your own secret key.
+  # config.secret_key = 'ab8a942770d9d8199572b390493c15337c90da5656368735edcfe12b05a9fca688d713769420c035ad8cbf8bff90e5607811c741c78a28438f31251cab8a1384'
```
Isso é importante! Dá uma lida e entenda o que isso faz!

ricardopacheco commented on config/locales/pt-BR.yml:
```diff
@@ -0,0 +1,246 @@
+pt-BR:
```
Excelente!

ricardopacheco commented on config/locales/devise.pt-BR.yml:
```diff
@@ -0,0 +1,65 @@
+# Traduções adicionais em https://github.com/plataformatec/devise/wiki/I18n
```
Excelente!

ricardopacheco commented on config/locales/devise.en.yml:
```diff
@@ -0,0 +1,65 @@
+# Additional translations at https://github.com/heartcombo/devise/wiki/I18n
```
Excelente!

ricardopacheco commented on config/routes.rb:
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  root to: 'municipes#index'
+  devise_for :users
```
Olha aí a questão de usar default. Talvez isso não esteja gerando mais rotas do que de fato vc usa?

Quero que analise isso e veja.

PedroPedrassolli replied:
Dei uma pesquisada, e parece que ele gera as rotas baseado no que está no model de User, que já arrumei. Acredito estar certinho.

ricardopacheco commented on config/routes.rb:
```diff
@@ -1,3 +1,5 @@
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  root to: 'municipes#index'
+  devise_for :users
+  resources :municipes, except: :destroy
```
Business rules on fire! <3

ricardopacheco commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,311 @@
+  # initial account confirmation) to be applied. Requires additional unconfirmed_email
+  # db field (see migrations). Until confirmed, new email is stored in
+  # unconfirmed_email column, and copied to email column on successful confirmation.
+  config.reconfirmable = true
```
Ué, enabled?

Mas você não habilitou o confirmable do devise, lembra? Apesar de ter gerado views e rotas, o model não dá preload desse módulo. Aqui você já teria problemas.

PedroPedrassolli replied:
Tem razão, já removí

ricardopacheco commented on package.json:
```diff
@@ -2,10 +2,16 @@
   "name": "municipes-app",
```
Gerenciamento de dependências de frontend feita no padrão correto! Excelente 👍

ricardopacheco commented on spec/models/municipe_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe Municipe, type: :model do
```
Conforme vc comentou está pending. Mas, já é possível testar! manda bala 🚀

PedroPedrassolli replied:
Com certeza! Logo mais sai um commit com validações do model!

ricardopacheco reviewed (changes requested):
Gostei bastante da forma pensada na aplicação. Muitos padrões corretos. De forma geral eu avalio super bem e parabéns!

Fiz alguns comentários que são pertinentes de mudança, por isso, a requisição de mudanças.

PedroPedrassolli committed:
fix(gemfile): adds versioning to gems i added before

PedroPedrassolli committed:
fix(municipe controller): changed so even if it doesnt find a municipe by id, it initializes one

PedroPedrassolli committed:
build(validations): adds cpf_cnpj, phonelib and model validations

PedroPedrassolli committed:
fix(messages/validations): adds flash messages to the screen and also new validations

i also added a CNS mask and some i18n

PedroPedrassolli committed:
build(decorator): installs draper and uses it to decorate the Municipes

PedroPedrassolli committed:
refactor(flash): refactors the way messages were being showed

n

PedroPedrassolli committed:
style(index table): changes positions for a "nicer" approach to the user

PedroPedrassolli committed:
style(i18n/views/decorator): adds i18n, removes useless views or adds new ones, adds decorator

PedroPedrassolli committed:
build(simplecov): installs SimpleCov

PedroPedrassolli committed:
refactor(rubocop): rubocop -a

n

PedroPedrassolli committed:
refactor(rubopcop): rubocop -A

PedroPedrassolli committed:
refactor(new array syntax): adds new array syntax

PedroPedrassolli committed:
perf(assets): removes precompile

also switches form-groups for col

PedroPedrassolli committed:
fix(application.scss): removes form-group from css

and from municipe form

PedroPedrassolli committed:
style(pt-br): change i18n

n

PedroPedrassolli committed:
fix(devise.rb): removes devise reconfirmable

PedroPedrassolli committed:
fix(application.rb): adds i18n fallbacks

italopires commented on app/decorators/municipe_decorator.rb:
```diff
@@ -0,0 +1,25 @@
+  end
+
+  def birthdate
+    object.birthdate.strftime('%d/%m/%Y')
```
Usar I18n para formatação da data, vai te dá o benefício de data por nacionalidade e padronização.

https://guides.rubyonrails.org/i18n.html#adding-date-time-formats

italopires commented on app/views/layouts/application.html.erb:
```diff
@@ -11,6 +11,8 @@
   </head>
 
   <body>
+    <%= render partial: "shared/navbar"%>
```
Lembrar sempre de usar aspas duplas apenas quando tiver interpolação de variáveis, nesse caso usaria apenas aspa simples. Essa uma boa prática, e o rubocop alarma muito isso

italopires commented on app/views/municipes/_form.html.erb:
```diff
@@ -0,0 +1,47 @@
+<%= form_for @municipe do |f| %>
+	<div class="container">
```
ajustaria a identação (2 espaços) nesse arquivo e nos demais.

Gostei que você está utilizando I18n!

italopires commented on app/views/municipes/_form.html.erb:
```diff
@@ -0,0 +1,47 @@
+<%= form_for @municipe do |f| %>
```
Dá uma olhada depois no simple_form, ele vai te ajudar a escrever menos html e vai te dá mais produtividade. Teu código vai ficar mais legível e manutenível.

https://github.com/heartcombo/simple_form

italopires commented on app/views/municipes/index.html.erb:
```diff
@@ -0,0 +1,31 @@
+	<table class="table table-striped table-bordered table-hover">
+		<thead>
+			<tr>
+				<th><%= I18n.t("activerecord.attributes.municipe.name")  %></th>
```
I18n, excelente!

italopires reviewed (changes requested):
Bom trabalho, continue assim.

PS: aguardando testes de código no próximo PR

PedroPedrassolli committed:
fix(devise mailer): removed it and its i18n's

PedroPedrassolli committed:
fix(application.rb): disables action text

PedroPedrassolli committed:
refactor(application.rb): makes it readable

PedroPedrassolli committed:
fix(decorator): fixes decorator

also added simple quotes and left the identation of every html 2 spaces per tab

gumiguelote commented on config/initializers/devise.rb:
```diff
@@ -24,7 +24,7 @@
   # note that it will be overwritten if you use your own mailer class
   # with default "from" parameter.
-  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
+  # config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
```
Código não usado, pode ser removido, lembre-se que se precisar vai existir no versionamento para consulta

gumiguelote commented on config/application.rb:
```diff
@@ -11,7 +11,7 @@
 require 'action_mailer/railtie'
 # require "action_mailbox/engine"
-require 'action_text/engine'
+# require 'action_text/engine'
```
Codigo não usado, remova, não comente. Deixa o código menos sujo.

gumiguelote commented on config/application.rb:
```diff
@@ -5,16 +5,16 @@
 require 'action_view/railtie'
 require 'action_cable/engine'
 require 'sprockets/railtie'
+# require "active_job/railtie"
```
Mesma regrinha, oq não for utilizar, remova do codigo.

gumiguelote reviewed (approved):
Muito bom! Da uma olhada nos comentários

gumiguelote commented on app/models/municipe.rb:
```diff
@@ -3,9 +3,23 @@ class Municipe < ApplicationRecord
+      errors.add(:cpf, I18n.t('errors.messages.invalid')) unless CPF.valid?(self.cpf)
+    end
+
+    def birthdate_data_range
```
![image](https://user-images.githubusercontent.com/25744569/127205916-9c36c1d4-a7c1-44ec-a107-daaa2612dcf9.png)

PedroPedrassolli committed:
fix(configs): removes useless code lines

PedroPedrassolli committed:
fix(municipe.rb): changes age validation to 120 maximum

PedroPedrassolli committed:
fix(rubocop.yml): adds excludes to rubocop and runs rubocop -A

PedroPedrassolli committed:
build(default_url): changes default_url on ApplicationController

so it's added dynamically to the mailer

ricardopacheco commented on Gemfile:
```diff
@@ -1,8 +1,18 @@
 
 ruby '3.0.1'
 
+# Draper
```
Style(opcional): O nome da gem já é por si só rs. Se vai comentar, coloca o propósito da gem, o que ela faz e o que vc pretende resolver com ela.

ricardopacheco commented on Gemfile:
```diff
@@ -28,9 +38,11 @@ gem 'jbuilder', '~> 2.7'
 # Reduces boot times through caching; required in config/boot.rb
```
Olha um exemplo aqui do que comentei acima!

ricardopacheco commented on Gemfile:
```diff
@@ -28,9 +38,11 @@ gem 'jbuilder', '~> 2.7'
 group :development, :test do
-  # Call 'byebug' anywhere in the code to stop execution and get a debugger console
-  gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
+  gem 'pry', '~> 0.13.1'
```
Style(opcional): Excelente escolha! Existe uma gem chamada `pry-rails` que já carrega o pry como dependência.

A vantagem é que ela tem alguns recursos exclusivos para rails apps. Consulte a documentação

ricardopacheco commented on app/channels/application_cable/channel.rb:
```diff
@@ -1,3 +1,5 @@
+# frozen_string_literal: true
+
 module ApplicationCable
```
Refactor(required): Sobre recursos utilizados. A app está utilizando websockets? Se está, como isso está feito?

Se não ... você já tem uma ideia do que fazer

ricardopacheco commented on app/channels/application_cable/connection.rb:
```diff
@@ -1,3 +1,5 @@
+# frozen_string_literal: true
```
Mesmo comentário acima.

ricardopacheco commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,9 @@
 class ApplicationController < ActionController::Base
+  before_action :authenticate_user!
+
+  def default_url_options
```
Refactor(required): Porque você acha que isso é necessário? Isso é utilizado pelo devise para recursos de rota.

Só que esses recursos são usados além dos controllers, mailers utilizam também (quando o devise envia email para recuperar senha por exemplo). Por isso, o local dessa configuração está impróprio. Deve ser configurado a nível de enviroment (ou seja, pode ser diferente em desenvolvimento, staging, produção).

Vou deixar o local correto pra vc pensar a respeito. E desculpe, `request.host_with_port` é CAGADO, me falta portugues pra adjetivar melhor isso. Pra não dizer que sou ruim, segue uma doc [aqui](https://guides.rubyonrails.org/action_mailer_basics.html#generating-urls-in-action-mailer-views)

ricardopacheco commented on app/controllers/municipes_controller.rb:
```diff
@@ -0,0 +1,54 @@
+  before_action :set_municipe, only: %i[show edit update]
+
+  def index
+    @municipes = Municipe.all.decorate
```
Nice! 👍

ricardopacheco commented on app/controllers/municipes_controller.rb:
```diff
@@ -0,0 +1,54 @@
+    if @municipe.save
+      saved_or_updated
+    else
+      failed('new')
```
Style(opcional): Interessante! Você pode tornar isso ainda maior, se colocar no lugar certo (pensa que outras classes de controllers podem querer o mesmo comportamento).

ricardopacheco commented on app/controllers/municipes_controller.rb:
```diff
@@ -0,0 +1,54 @@
+    if @municipe.update(municipe_params)
+      saved_or_updated
+    else
+      failed('edit')
```
Mesmo comentário acima

ricardopacheco commented on app/controllers/municipes_controller.rb:
```diff
@@ -0,0 +1,54 @@
+
+  def update
+    if @municipe.update(municipe_params)
+      saved_or_updated
```
Mesmo comentário acima.

ricardopacheco commented on app/controllers/municipes_controller.rb:
```diff
@@ -0,0 +1,54 @@
+  private
+
+  def set_municipe
+    @municipe = Municipe.find_or_initialize_by(id: params[:id])
```
Porque find_or_initialize_by ?

PedroPedrassolli replied:
Então, deixei find_or_initialize_by para quando ele não achar um municipe com certo id ele já inicializar um, aí a pessoa pode mexer ainda na página ao invés de cair em um 404, só não tenho certeza ainda se é a melhor coisa a se fazer

ricardopacheco commented on app/controllers/municipes_controller.rb:
```diff
@@ -0,0 +1,54 @@
+  end
+
+  def failed(action)
+    flash.now[:danger] = @municipe.errors.full_messages.to_sentence
```
Você está usando decorators, lembra?

ricardopacheco commented on app/decorators/municipe_decorator.rb:
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+class MunicipeDecorator < Draper::Decorator
```
Excepcional! 💯

ricardopacheco commented on app/decorators/municipe_decorator.rb:
```diff
@@ -0,0 +1,25 @@
+  end
+
+  def birthdate
+    I18n.l object.birthdate
```
Aqui sem parenteses, na linha 16 e 18 com parênteses. Com ou sem, está correto! Mas, que tal um padrão de cofificação? ;)

ricardopacheco commented on app/mailers/application_mailer.rb:
```diff
@@ -1,3 +1,5 @@
+# frozen_string_literal: true
+
 class ApplicationMailer < ActionMailer::Base
   default from: 'from@example.com'
```
Há uma melhor forma de configurar isso. Se eu quisesse em desenvolvimento usar o domain om30.local e, em produção om30.com.br, como isso deveria ser configurado?

ricardopacheco commented on app/models/municipe.rb:
```diff
@@ -0,0 +1,32 @@
+
+  validates :name, :cpf, :cns, :email, :birthdate, :phone, :photo, presence: true
+  validates :status, inclusion: { in: [true, false] }
+  validates :name, format: { with: /([\w\-']{2,})(\s+)([\w\-']{2,})/ }, if: :name
```
if name what? rs

Talvez vc queira ser mais explícito em uma validação. `if 'name.present?'`. E, mesmo com esse exemplo, tem um jeito mais elegante pra usar o if. Opcional, mas, tenta descobrir qual.

ricardopacheco commented on app/models/municipe.rb:
```diff
@@ -0,0 +1,32 @@
+  validates :cpf, :cns, numericality: { only_integer: true }
+  validates :phone, length: { is: 11 }
+  validates :phone, phone: true, if: :phone
+  validates_format_of :email, with: Devise.email_regexp, if: :email
```
Excelente opção! Usou um método que já é amplamente utilizado (logo, testado). Mandou bem

ricardopacheco commented on app/models/municipe.rb:
```diff
@@ -0,0 +1,32 @@
+  end
+
+  def birthdate_data_range
+    unless ((120.years.ago)..(1.day.ago)).include?(birthdate)
```
Style(opcional): que tal? `return true if ((120.years.ago)..(1.day.ago)).include?(birthday)`

```
def birthdate_data_range
  return true if ((120.years.ago)..(1.day.ago)).include?(birthday)
  errors.add(:birthdate,I18n.t('errors.messages.invalid'))
end
```

ricardopacheco commented on app/models/user.rb:
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+class User < ApplicationRecord
+  devise :database_authenticatable, :registerable, :rememberable, :validatable
```
Leve e enxuto! <3

ricardopacheco commented on app/views/devise/registrations/edit.html.erb:
```diff
@@ -0,0 +1,37 @@
+    <div class="col">
+      <%= f.label :password %>
+        <i>
+          <% if @minimum_password_length %>
```
Que tal um helper maroto?

ricardopacheco commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,29 @@
+      <%= f.label :password %>
+        <i>
+          (
+          <% if @minimum_password_length %>
```
Mesmo comentário acima

ricardopacheco commented on app/views/devise/shared/_error_messages.erb:
```diff
@@ -0,0 +1,6 @@
+<% if resource.errors.present? %>
```
Nice partial! 👍

ricardopacheco commented on app/views/layouts/application.html.erb:
```diff
@@ -11,6 +11,8 @@
   </head>
 
   <body>
+    <%= render partial: 'shared/navbar'%>
```
Syntax: space after final string

ricardopacheco commented on app/views/layouts/application.html.erb:
```diff
@@ -11,6 +11,8 @@
 
   <body>
+    <%= render partial: 'shared/navbar'%>
+    <%= render partial: 'shared/error_messages'%>
```
Same

ricardopacheco commented on config/application.rb:
```diff
@@ -1,19 +1,17 @@
-require "action_view/railtie"
-require "action_cable/engine"
-require "sprockets/railtie"
-# require "rails/test_unit/railtie"
```
Esse é um dos RARÍSSIMOS casos que não recomendo a remoção, apenas comentar desabilitado. Imagina que daqui a alguns anos, o projeto precisa utilizar action text ou action mailbox? Decisões sobre tools podem mudar conforme a necessidade.

Vc fez correto! Porém, lembre-se disso.

ricardopacheco commented on config/application.rb:
```diff
@@ -24,6 +22,12 @@ class Application < Rails::Application
 
+    config.i18n.default_locale = :"pt-BR"
+    config.i18n.available_locales = %i[pt-BR en]
+    config.i18n.fallbacks = %i[pt-BR en]
```
🥇  Excelente

ricardopacheco commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,309 @@
+  # Email regex used to validate email formats. It simply asserts that
+  # one (and only one) @ exists in the given string. This is mainly
+  # to give user feedback and not to assert the e-mail validity.
+  config.email_regexp = /\A[^@\s]+@[^@\s]+\z/
```
Curiosidade: Devise.email_regexp vem desse carinha ;)

ricardopacheco commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,309 @@
+
+  # When set to false, does not sign a user in automatically after their password is
+  # changed. Defaults to true, so a user is signed in automatically after changing a password.
+  # config.sign_in_after_change_password = true
```
Isso pode ser importante pra um módulo que vc está usando. A dica é pra entender o caso e ver se necessita ou não.

ricardopacheco commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,309 @@
+  # Time interval you can reset your password with a reset password key.
+  # Don't put a too small interval or your users won't have the time to
+  # change their passwords.
+  config.reset_password_within = 6.hours
```
Essa configuração em produção poderia afetar seus usuários. Como?

Tente entender.

ricardopacheco commented on config/initializers/phonelib.rb:
```diff
@@ -0,0 +1,3 @@
+# frozen_string_literal: true
+
+Phonelib.default_country = 'BR'
```
Initializer, gostei! 👍

ricardopacheco commented on coverage/.last_run.json:
```diff
@@ -0,0 +1,5 @@
+{
```
Isso é uma crítica severa: Deixar commitável o relatório de coverage de código é um descuido desprestencioso, porém, faltou atenção na hora do commit. Relatórios podem mudar a cada alteração, isso é como arquivo de log. Você quer ter, mas, não pra distribuir como código entre o time!

Crítica: TENHA MAIS ATENÇÃO AOS COMMITS!

ricardopacheco commented on db/migrate/20210720141304_create_municipe.rb:
```diff
@@ -0,0 +1,17 @@
+  def change
+    create_table :municipes do |t|
+      t.string :name
+      t.integer :cpf
```
Refactor: Por aqui eu já sei que sua validação uniqueness não irá funcionar. Afinal de contas, quem garante a validação dos dados, é o banco de dados ou o ORM (activerecord) ?

Leitura RECOMENDADA: https://thoughtbot.com/blog/the-perils-of-uniqueness-validations

ricardopacheco commented on db/migrate/20210720141304_create_municipe.rb:
```diff
@@ -0,0 +1,17 @@
+    create_table :municipes do |t|
+      t.string :name
+      t.integer :cpf
+      t.integer :cns
```
Mesmo comentário acima

ricardopacheco commented on db/migrate/20210720141304_create_municipe.rb:
```diff
@@ -0,0 +1,17 @@
+      t.string :name
+      t.integer :cpf
+      t.integer :cns
+      t.string :email
```
Mesmo comentário acima

ricardopacheco commented on db/migrate/20210720141304_create_municipe.rb:
```diff
@@ -0,0 +1,17 @@
+      t.integer :cns
+      t.string :email
+      t.date :birthdate
+      t.integer :phone
```
Mesmo comentário acima

ricardopacheco commented on spec/decorators/municipe_decorator_spec.rb:
```diff
@@ -0,0 +1,6 @@
+
+require 'rails_helper'
+
+RSpec.describe MunicipeDecorator do
```
tinha aproveitado e já escrito as  specs para os decorators. Mas, já deixou pendente, isso já ganha pontos o/

ricardopacheco reviewed (changes requested):
Gostei do PR! Consegui perceber uma pequena evolução em seu código. Fiz muitos comentários pertinentes, porém opcionais.

Porém, alguns pontos estão críticos e inclusive lhe causaria problemas de deployment em plataformas de saas por isso devem ser refatorados.

italopires commented on Rakefile:
```diff
@@ -1,6 +1,8 @@
 # for example lib/tasks/capistrano.rake, and they will automatically be available to Rake.
 
-require_relative "config/application"
+require_relative 'config/application'
```
Excelente as aspas simples!
